crates/environ/src/module.rs

```diff
@@ -192,88 +170,71 @@ impl ModuleTranslation<'_> {
     /// initialization then this won't change anything. Otherwise if it is
     /// compatible then the `memory_initialization` field will be updated.
     pub fn try_paged_init(&mut self) {
-        let initializers = match &self.module.memory_initialization {
-            MemoryInitialization::Segmented(list) => list,
-            MemoryInitialization::Paged { .. } => return,
-        };
-        let page_size = u64::from(WASM_PAGE_SIZE);
-        let num_defined_memories =
-            self.module.memory_plans.len() - self.module.num_imported_memories;
-        let mut out_of_bounds = false;
+        // This method only attempts to transform a a `Segmented` memory init
+        // into a `Paged` one, no other state.
+        match &self.module.memory_initialization {
+            MemoryInitialization::Segmented(_) => {}
+            _ => return,
+        }
 
         // Initially all memories start out as all zeros, represented with a
         // lack of entries in the `BTreeMap` here. The map indexes byte offset
         // (which is always wasm-page-aligned) to the contents of the page, with
         // missing entries implicitly as all zeros.
-        let mut page_contents = PrimaryMap::with_capacity(num_defined_memories);
-        for _ in 0..num_defined_memories {
+        let mut page_contents = PrimaryMap::with_capacity(self.module.memory_plans.len());
+        for _ in 0..self.module.memory_plans.len() {
             page_contents.push(BTreeMap::new());
         }
 
-        assert_eq!(initializers.len(), self.data.len());
-        for (initializer, data) in initializers.iter().zip(&self.data) {
-            let memory_index = match (
-                self.module.defined_memory_index(initializer.memory_index),
-                initializer.base.is_some(),
-            ) {
-                (None, _) | (_, true) => {
-                    // If the initializer references an imported memory or uses a global base,
-                    // the complete set of segments will need to be processed at module instantiation
-                    return;
+        // Perform a "dry run" of memory initialization which will fail if we
+        // can't switch to paged initialization. When data is written it's
+        // transformed into the representation of `page_contents`.
+        let mut data = self.data.iter();
+        let result = self.module.memory_initialization.init_memory(
+            |memory_index| self.module.memory_plans[memory_index].memory.minimum,
```

Comment:
I'm not opposed to such an idiom but personally I feel it's more motivation for a different API (it's a pain otherwise to ensure all call-sites have comments like this), so I brought back the `InitMemory` enum as a way to encapsulate the first two arguments here so there's only one closure-as-argument.

Reply:
Wow, nice, the enum-of-named-args is fantastic. Thanks!